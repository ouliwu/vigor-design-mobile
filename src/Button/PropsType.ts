/**
 * Button properties
 */
export interface ButtonPropsType {
  /** 按钮样式 */
  type?: 'default' | 'primary' | 'warn';
  /** 是否禁用按钮 */
  disabled?: boolean;
  /** 按钮点触事件 */
  onPress?: (e?: React.SyntheticEvent) => void;
  /** 按钮是否显示加载中 */
  loading?: boolean;
  /** 额外的class名 */
  className?: string;
  /** 按钮active状态下额外的class名 */
  activeClassName?: string;
}
