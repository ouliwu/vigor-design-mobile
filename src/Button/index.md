---
title: Button
nav:
  title: 组件
  path: /components
---

## 基础使用

### type

通过 type 属性切换 Button 组件样式

```tsx
import React from 'react';
import { Button } from 'vigor-design-mobile';

export default () => {
  return (
    <>
      <Button>default</Button>
      <br />
      <Button type="primary">primary</Button>
      <br />
      <Button type="warn">warn</Button>
    </>
  );
};
```

### disabled

通过 disabled 属性设置按钮是否禁用

```tsx
import React from 'react';
import { Button } from 'vigor-design-mobile';

export default () => {
  return (
    <>
      <Button
        disabled={true}
        onPress={() => {
          console.log(1);
        }}
      >
        default
      </Button>
      <br />
      <Button disabled={true} type="primary">
        primary
      </Button>
      <br />
      <Button disabled={true} type="warn">
        warn
      </Button>
    </>
  );
};
```

<API></API>
