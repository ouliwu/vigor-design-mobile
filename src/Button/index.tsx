import React, { useState } from 'react';
import './index.scss';
import { ButtonPropsType } from './PropsType';
import Touchable from '@ant-design/mobile-touchable';
import classNames from 'classnames';

const prefixCls = 'h-button';

const Button: React.FC<ButtonPropsType> = props => {
  const {
    type,
    disabled,
    onPress,
    loading,
    className,
    activeClassName,
    children,
  } = props;
  const [cls, setCls] = useState({
    [prefixCls]: true,
    [`${prefixCls}-active`]: false,
    [activeClassName!]: false,
  });
  const innerCls = classNames({
    ...cls,
    [`${prefixCls}-${type}`]: true,
    [className!]: !!className,
    [`${prefixCls}-disabled`]: disabled,
    [`${prefixCls}-loading`]: loading,
  });
  return (
    <Touchable
      disabled={loading || disabled}
      onPressIn={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: true,
          [activeClassName!]: true,
        });
      }}
      onPress={e => {
        onPress && onPress(e);
      }}
      onPressOut={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: false,
          [activeClassName!]: false,
        });
      }}
    >
      <div className={innerCls}>
        <div className={`${prefixCls}-loading-container`}>
          <div className={`${prefixCls}-loading-main`} />
        </div>
        <div className={`${prefixCls}-text-container`}>
          <span>{children}</span>
        </div>
      </div>
    </Touchable>
  );
};

Button.defaultProps = {
  type: 'default',
  disabled: false,
  loading: false,
  className: '',
  activeClassName: '',
};

export default Button;
