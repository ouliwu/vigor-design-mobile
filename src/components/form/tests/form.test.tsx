import React from 'react'
import { fireEvent, render, waitFor } from 'testing'
import Form from '..'
import Input from '../../input'
import Button from '../../button'
import Checkbox from '../../checkbox'

const classPrefix = `adm-form`

function $$(className: string) {
  return document.body.querySelectorAll(className)
}

const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {})

describe('Form', () => {
  afterEach(() => {
    warnSpy.mockReset()
  })

  afterAll(() => {
    warnSpy.mockRestore()
  })

  test('basic usage', async () => {
    const fn = jest.fn()

    const { getByText, getByLabelText } = render(
      <Form
        onFinish={fn}
        footer={
          <Button block type='submit' color='primary' size='large'>
            submit
          </Button>
        }
      >
        <Form.Header>基础用法</Form.Header>
        <Form.Item name='name' label='name' rules={[{ required: true }]}>
          <Input placeholder='请输入姓名' />
        </Form.Item>
        <Form.Item name='address' label='address' help='address detail'>
          <Input placeholder='请输入地址' />
        </Form.Item>
      </Form>
    )

    console.error = jest.fn()

    fireEvent.click(getByText('submit'))
    await waitFor(() => {
      expect($$(`.${classPrefix}-item-feedback-error`).length).toBeTruthy()
    })

    expect($$(`.${classPrefix}-item-feedback-error`).length).toBeTruthy()
    expect($$(`.${classPrefix}-item-has-error`).length).toBeTruthy()
    expect(console.error).toBeCalledTimes(0)

    fireEvent.change(getByLabelText(/name/i), { target: { value: 'name' } })
    fireEvent.change(getByLabelText(/address/i), {
      target: { value: 'address' },
    })

    fireEvent.click(getByText('submit'))
    expect(console.error).toBeCalledTimes(0)
    await waitFor(() => {
      expect(fn.mock.calls[0][0]).toEqual({ name: 'name', address: 'address' })
    })
  })

  test('renders with horizontal layout', async () => {
    render(
      <Form layout='horizontal'>
        <Form.Item data-testid='form-item'>
          <Input />
        </Form.Item>
      </Form>
    )

    expect($$(`.${classPrefix}-item`)[0]).toHaveClass(
      `${classPrefix}-item-horizontal`
    )
  })

  test('custom field update logic', async () => {
    const { getByText, getByTestId, getByLabelText } = render(
      <Form>
        <Form.Item name='a' label='A'>
          <Input />
        </Form.Item>
        <Form.Item name='b' label='B' required>
          <Checkbox.Group>
            <Checkbox value='1'>选项1</Checkbox>
            <Checkbox value='2'>选项2</Checkbox>
          </Checkbox.Group>
        </Form.Item>
        <Form.Item
          label='表单联动-字段B'
          shouldUpdate={(prevValues, curValues) => {
            return prevValues.b !== curValues.b
          }}
        >
          {({ getFieldValue }) => (
            <div data-testid='res'>{JSON.stringify(getFieldValue('b'))}</div>
          )}
        </Form.Item>
      </Form>
    )

    fireEvent.click(getByText('选项1'))
    expect(getByTestId('res')).toHaveTextContent('["1"]')

    fireEvent.change(getByLabelText(/A/), { target: { value: 'aaa' } })
    expect(getByTestId('res')).toHaveTextContent('["1"]')
  })

  test('return same form instance', async () => {
    const instances = new Set()

    const App = () => {
      const [form] = Form.useForm()
      instances.add(form)
      const [, forceUpdate] = React.useState({})
      return (
        <button
          type='button'
          onClick={() => {
            forceUpdate({})
          }}
        >
          Refresh
        </button>
      )
    }

    const { getByText } = render(<App />)

    for (let i = 0; i < 5; i += 1) {
      fireEvent.click(getByText('Refresh'))
    }

    expect(instances.size).toEqual(1)
  })

  test('`messageVariables` support validate', async () => {
    const { getByTestId } = render(
      <Form
        data-testid='form'
        validateMessages={{ required: "'${name}' is required" }}
      >
        <Form.Item name='test' rules={[{ required: true }]}>
          <Input />
        </Form.Item>
      </Form>
    )

    fireEvent.submit(getByTestId('form'))

    await waitFor(() => {
      expect(getByTestId('form')).toHaveTextContent(`'test' is required`)
    })
  })

  test("`shouldUpdate` shouldn't work with render props", async () => {
    render(
      <Form>
        <Form.Item>{() => null}</Form.Item>
      </Form>
    )

    expect(warnSpy).toHaveBeenCalledWith(
      '[vigor-design-mobile: Form.Item] `children` of render props only work with `shouldUpdate` or `dependencies`.'
    )
  })

  test("`shouldUpdate` shouldn't work with `dependencies`", async () => {
    render(
      <Form>
        <Form.Item shouldUpdate dependencies={[]}>
          {() => null}
        </Form.Item>
      </Form>
    )

    expect(warnSpy).toHaveBeenCalledWith(
      "[vigor-design-mobile: Form.Item] `shouldUpdate` and `dependencies` shouldn't be used together."
    )
  })

  test("`name` shouldn't work with render props", async () => {
    render(
      <Form>
        <Form.Item name='test' shouldUpdate>
          {() => null}
        </Form.Item>
      </Form>
    )

    expect(warnSpy).toHaveBeenCalledWith(
      "[vigor-design-mobile: Form.Item] Do not use `name` with `children` of render props since it's not a field."
    )
  })

  test('warning when use `dependencies` but `name` is empty & children is not a render props', async () => {
    render(
      <Form>
        <Form.Item dependencies={[]}>text</Form.Item>
      </Form>
    )

    expect(warnSpy).toHaveBeenCalledWith(
      '[vigor-design-mobile: Form.Item] Must set `name` or use render props when `dependencies` is set.'
    )
  })

  test('warning with `defaultValue`', async () => {
    render(
      <Form>
        <Form.Item name='test'>
          <Input defaultValue='test' />
        </Form.Item>
      </Form>
    )

    expect(warnSpy).toHaveBeenCalledWith(
      '[vigor-design-mobile: Form.Item] `defaultValue` will not work on controlled Field. You should use `initialValues` of Form instead.'
    )
  })

  test('warning when use `name` but children is not validate element', async () => {
    render(
      <Form>
        <Form.Item name='test'>text</Form.Item>
      </Form>
    )

    expect(warnSpy).toHaveBeenCalledWith(
      '[vigor-design-mobile: Form.Item] `name` is only used for validate React element. If you are using Form.Item as layout display, please remove `name` instead.'
    )
  })

  test('warningOnly validate', async () => {
    const fn = jest.fn()
    const { getByTestId } = render(
      <Form data-testid='form' onFinish={fn}>
        <Form.Item
          name='email'
          label='邮箱'
          rules={[
            { required: true },
            { type: 'string', min: 6 },
            { type: 'email', warningOnly: true },
          ]}
        >
          <Input placeholder='请输入邮箱' />
        </Form.Item>
      </Form>
    )

    fireEvent.change($$('input')[0], { target: { value: 'aaaaaa' } })

    await waitFor(() => {
      expect($$(`.${classPrefix}-item-feedback-error`).length).not.toBeTruthy()
      expect($$(`.${classPrefix}-item-feedback-warning`).length).toBeTruthy()
    })

    fireEvent.submit(getByTestId('form'))

    await waitFor(() => {
      expect(fn).toBeCalledTimes(1)
    })
  })

  describe('Form.Item', () => {
    test('noStyle', async () => {
      const onChange = jest.fn()
      const { container } = render(
        <Form>
          <Form.Item noStyle name='test'>
            <Input onChange={onChange} />
          </Form.Item>
        </Form>
      )

      expect(container).toMatchSnapshot()

      fireEvent.change($$('input')[0], { target: { value: 'test' } })
      expect(onChange).toHaveBeenCalled()
    })

    test('hidden', async () => {
      const onChange = jest.fn()
      const { container } = render(
        <Form>
          <Form.Item hidden name='test'>
            <Input onChange={onChange} />
          </Form.Item>
        </Form>
      )

      expect(container).toMatchSnapshot()

      fireEvent.change($$('input')[0], { target: { value: 'test' } })
      expect(onChange).toHaveBeenCalled()
    })
  })
})
