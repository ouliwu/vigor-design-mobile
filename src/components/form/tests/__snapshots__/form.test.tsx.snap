// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Form.Item hidden 1`] = `
<div>
  <form
    class="adm-form"
  >
    <div
      class="adm-list adm-list-default"
    >
      <div
        class="adm-list-body"
      >
        <div
          class="adm-list-body-inner"
        >
          <div
            class="adm-list-item adm-form-item adm-form-item-vertical adm-form-item-hidden"
          >
            <div
              class="adm-list-item-content"
            >
              <div
                class="adm-list-item-content-main"
              >
                <div
                  class="adm-form-item-child adm-form-item-child-position-normal"
                >
                  <div
                    class="adm-form-item-child-inner"
                  >
                    <div
                      class="adm-input"
                    >
                      <input
                        class="adm-input-element"
                        id="test"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="adm-list-item-description"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Form Form.Item noStyle 1`] = `
<div>
  <form
    class="adm-form"
  >
    <div
      class="adm-list adm-list-default"
    >
      <div
        class="adm-list-body"
      >
        <div
          class="adm-list-body-inner"
        >
          <div
            class="adm-input"
          >
            <input
              class="adm-input-element"
              id="test"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
